### PR TITLE
Update small obb candidates selection

### PR DESCRIPTION
**Pull request by lmycross**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves rotated-box training stability by hardening angle-loss math and making target assignment more robust for tiny ground-truth boxes 🧠🔧

### 📊 Key Changes
- Added small epsilon values in rotated angle loss to prevent invalid `log(w/h)` when width/height are near zero, avoiding NaNs/Infs 🛡️
- Updated rotated target assignment (`select_candidates_in_gts`) from `@staticmethod` to an instance method so it can use `self.stride` 🧩
- Introduced stride-aware clamping: if a GT rotated box has very small `w/h` (below a stride threshold), its size is bumped to a stride value during candidate selection to avoid degenerate geometry 📏

### 🎯 Purpose & Impact
- Reduces training crashes and instability for oriented/rotated bounding box tasks by preventing numerical edge cases ✅
- Improves positive-anchor selection reliability for extremely small rotated objects, which can otherwise be missed or produce noisy assignments 🎯
- Leads to more consistent training behavior and potentially better convergence/accuracy on datasets with tiny rotated targets (e.g., aerial imagery, document/OCR, industrial parts) 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lmycross, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist (this is automated, and an engineer will assist soon) 🧩✅

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/tal.py [hunk 1]
```diff
@@ -357,19 +357,21 @@ def iou_calculation(self, gt_bboxes, pd_bboxes):
...
         Args:
             xy_centers (torch.Tensor): Anchor center coordinates with shape (h*w, 2).
             gt_bboxes (torch.Tensor): Ground truth bounding boxes with shape (b, n_boxes, 5).
             mask_gt (torch.Tensor): Mask for valid ground truth boxes with shape (b, n_boxes, 1).
-            stride (list[int]): List of stride values for each feature map level.
 
         Returns:
             (torch.Tensor): Boolean mask of positive anchors with shape (b, n_boxes, h*w).
         """
+        wh_mask = gt_bboxes[..., 2:4] < self.stride[0]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/tal.py:371]: 💡 **MEDIUM**: `gt_bboxes` is modified in-place, which may leak the clamped sizes to callers that reuse the same tensor later in the assignment process. If `gt_bboxes` is expected to remain unchanged, this can cause subtle downstream issues. Consider operating on a cloned tensor for the candidate selection step.

File: ultralytics/utils/tal.py [hunk 2]
```diff
@@ -357,19 +357,21 @@ def iou_calculation(self, gt_bboxes, pd_bboxes):
...
+        stride_val = torch.tensor(self.stride[1], dtype=gt_bboxes.dtype, device=gt_bboxes.device)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/tal.py:372]: 💡 **MEDIUM**: `self.stride[1]` assumes at least two stride levels. If a model is configured with a single stride (or a different stride list length), this will raise an IndexError during training. Consider using a safe index (e.g., first stride) or guarding for length.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good targeted changes for stability. I found a few potential issues in the new candidate-selection logic: possible IndexError if only one stride is defined, in-place modification of `gt_bboxes`, and inconsistent stride selection for the clamp threshold vs. replacement value. Addressing these should make the change safer across configurations.

💬 Posted 2 inline comments

**Review by Laughing-q — APPROVED** [review]
@lmycross Looks good!

**Comment by UltralyticsAssistant** [comment]
Merged — huge thanks @lmycross (and @Laughing-q for the support and contributions)! 🙌  

“**Perseverance is not a long race; it is many short races one after the other.**” — **Walter Elliot**  
This PR is a perfect example of that mindset: tightening up the rotated angle-loss math with small epsilons to eliminate NaNs/Infs, and making stride-aware target assignment more resilient for tiny rotated GTs. The result is sturdier oriented-box training and more reliable positives on the hard edge cases (aerial, OCR, industrial tiny parts) — exactly the kind of robustness that helps the whole community ship better models.

Appreciate you both helping make Ultralytics training more dependable for everyone.
